neo/Consensus/ConsensusService.cs

```diff
@@ -83,49 +95,80 @@ private void ChangeTimer(TimeSpan delay)
             }, ActorRefs.NoSender);
         }
 
-        private void CheckExpectedView(byte view_number)
+        private void CheckCommits()
         {
-            if (context.ViewNumber == view_number) return;
-            if (context.ExpectedView.Count(p => p == view_number) >= context.M)
+            if (context.CommitPayloads.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
             {
-                InitializeConsensus(view_number);
+                Block block = context.CreateBlock();
+                Log($"relay block: {block.Hash}");
+                localNode.Tell(new LocalNode.Relay { Inventory = block });
+                context.State |= ConsensusState.BlockSent;
             }
         }
 
-        private void CheckSignatures()
+        private void CheckExpectedView(byte viewNumber)
         {
-            if (context.Signatures.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
+            if (context.ViewNumber == viewNumber) return;
+            if (context.ChangeViewPayloads.Count(p => p != null && p.GetDeserializedMessage<ChangeView>().NewViewNumber == viewNumber) >= context.M)
             {
-                Block block = context.CreateBlock();
-                Log($"relay block: {block.Hash}");
-                localNode.Tell(new LocalNode.Relay { Inventory = block });
-                context.State |= ConsensusState.BlockSent;
+                ChangeView message = context.ChangeViewPayloads[context.MyIndex]?.GetDeserializedMessage<ChangeView>();
+                if (message is null || message.NewViewNumber < viewNumber)
+                    localNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeChangeView(viewNumber) });
+                InitializeConsensus(viewNumber);
             }
         }
 
-        private void InitializeConsensus(byte view_number)
+        private void CheckPreparations()
         {
-            if (view_number == 0)
-                context.Reset();
-            else
-                context.ChangeView(view_number);
+            if (context.PreparationPayloads.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
+            {
+                ConsensusPayload payload = context.MakeCommit();
+                Log($"send commit");
+                context.State |= ConsensusState.CommitSent;
+                context.Save(store);
+                localNode.Tell(new LocalNode.SendDirectly { Inventory = payload });
```

Comment:
I am not fully convinced about this order here.
I fell that we could send this payload before `context.Save(store);`.

* Does localNode.Tell happens asynchronously?
* Save happens asynchronously, right? But does it move to localNode.Tell just when the context is saved?
* Why we would have more change to crash after or during broadcast?

Reply:
Saving the context is synchronous now. In this design. It is ensuring there is no way the node will not be committed while sending a commit message. If the order were opposite and he node crashed before persisting the context then it could be possible it could restart and commit in a different view. This could lead to forked blocks which is unacceptable.

Reply:
I got it now, this ensures a local level of safety. Thanks,